Extensions: Add extension redux state

coderkevin opened the pull request:
***Notice***
This PR is being deferred in favor of a more static approach for now (#11728). However, the dynamic approach will need to be revisited at some point in the future. Issue #12582 was created to track the future implementation of this enhancement.

Note: I'd like to keep this PR open until it is implemented, so as to not lose the code that exists on this branch until it has been successfully re-implemented.
***


This PR adds redux state support to extensions by way of enabling dynamic addition of reducers to the store.

It also creates a `withReducer()` Higher Order Component which ties a reducer to a React component for its lifecycle (i.e. reducer is added upon mount, removed upon dismount) This allows the reducer to be present for the duration of the use of the extension, but not be present when the extension is not in use.

Note: This triggers a console error log from redux's
combineReducers() function. This is caused by the persistence
of top-level state keys brought in to the initial state before
the extension has added its reducer. It's not a problem, but
should become a discussion point for what's next. (Issue logged as #11455)

To Test:
1. Run `make run`
2. Visit `http://calypso.localhost:3000/hello-dolly`
3. Verify that the lyrics show up in bold.
4. Click 'Next' to see the next lyric.

Still to be done:

* [x] Add reducer support.
* [x] Add Hello Dolly reducer.
* [x] Add actions to Hello Dolly example.
* [x] Add Lyric UI component to Hello Dolly example which reads from Redux state.

### Pre-Merge Checklist ###

 - [ ] Tests in Docker
 - [ ] Clear the browser caches `localStorage.clear(); indexedDB.deleteDatabase( 'calypso' )`
 - [ ] Sign in outside of the proxy and create a user and a site
 - [ ] Write and publish a post
 - [ ] Test on staging after merge and before deploy

dmsnell commented on client/extensions/hello-dolly/index.js:
```diff
@@ -14,16 +14,33 @@ import Card from 'components/card';
+ * Extension init
+ */
+// TODO: Maybe make this part of the extension load/unload code for all extensions?
+addReducer( 'hello-dolly', reducer );
```
since our naming conventions are built around `camelCase` it seems like that might be more appropriate for `helloDolly` than `kebab-case` is

coderkevin replied:
Thanks!

dmsnell commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,63 @@
+	switch ( action.type ) {
+		case HELLO_DOLLY_NEXT_LYRIC:
+			advanceToNextLyric( state );
+		default:
```
just a note that we tend to discourage fall-through logic due to the ease with which bugs can slip in unnoticed…

coderkevin replied:
Gotcha, thanks!

dmsnell commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,63 @@
+
+function advanceToNextLyric( state ) {
+	const { lyricIndex } = state;
+	const newIndex = ( lyricIndex < lyrics.length - 1 ? lyricIndex + 1 : 0 );
```
no need for a ternary! there's a math utility for that - modulus

```js
const nextIndex = ( lyricIndex + 1 ) % lyrics.length;
```

coderkevin replied:
Hmm, I haven't done it that way before. Seems a bit clever, but I can go for it.

dmsnell replied:
:+1: may be unfamiliar, but [modulo arithmetic](https://en.wikipedia.org/wiki/Modular_arithmetic) is actually a very common pattern well established. it's great because it's one of those cases where the result of evaluating the expression is "always valid" and we need no conditional branching (at least in the code we're writing).

coderkevin replied:
I've used modulus plenty of times for other things, but never had the idea to use it on the length of an array. Thanks for the tip!

dmsnell commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,63 @@
+];
+
+const initialState = {
+	lyric: lyrics[ 0 ],
```
it would be nice for example code to illustrate the concept of storing the minimum necessary state to represent the app. in this case, why do we need to store the lyric? do we plan on adding lyrics at runtime? if not, why not just store the index and allow some selector to pull out the text of the referenced lyrics?

keeping state access indirect will also give us flexibility as we grow, so might be good for example code

coderkevin replied:
Yeah, I suppose there's an opportunity to show how a selector can work in an extension. I'll give it a try.

dmsnell commented on client/state/index.js:
```diff
@@ -59,10 +60,12 @@ import ui from './ui/reducer';
 import users from './users/reducer';
 import wordads from './wordads/reducer';
 
+const log = debug( 'calypso:state:reducers' );
```
just for your information, usually we're importing something like `debugFactory` in Calypso and then creating `const debug = debugFactory( … )`

coderkevin replied:
Okay, I'll look up that example then. I saw both and wasn't sure which to do.

dmsnell commented on client/state/index.js:
```diff
@@ -111,7 +114,40 @@ export const reducer = combineReducers( {
+
+export function reducerExists( name ) {
+	return ( reducers[ name ] !== undefined );
+}
```
these helpers almost certainly belong in their own file/directory to separate them from the core state code in Calypso

coderkevin replied:
Given that they modify global state, I guess we'd have to export the reducers variable? But that's exactly the global we don't want to mutate directly, right?

I was thinking of going with a class to encapsulate this a bit, but if you have another suggestion, I'm open to it.

dmsnell replied:
what might use this outside of the module itself?

if this is the only module needing to use this function then we could easily allow it to accept the reducer object…

```js
export const reducerExists = ( reducers, name ) => reducers[ name ] !== undefined;
```

of course at this point we recognize a common code pattern we might be able to reuse…

```js
if ( has( reducers, name ) ) {
	…
}
```

coderkevin replied:
`reducerExists()` was only really used in one place here, so I just omitted the function altogether for now. If we need this kind of function in the future, we can easily create it again.

dmsnell commented on client/state/index.js:
```diff
@@ -111,7 +114,40 @@ export const reducer = combineReducers( {
+
+export function addReducer( name, reducerFunc ) {
+	if ( reducerExists( name ) ) {
+		throw new Error( 'addReducer(): name "' + name + '" already in use' );
```
a good place to use template literals!

```js
`addReducer(): name ${ name } already in use`
```

coderkevin replied:
👍

dmsnell commented on client/state/index.js:
```diff
@@ -111,7 +114,40 @@ export const reducer = combineReducers( {
+		return;
+	}
+
+	reducers = remainingReducers;
```
these mutating assignment lines scare me! I wonder if there's a way to modify the chain of reducers that doesn't involve shared mutable state…

coderkevin replied:
Yeah, I'm not really a fan of this either. I was going for minimal changes.
However, if others agree, I'll add some more here to avoid mutating globals directly like this.

dmsnell replied:
I would take extra time and thought to remove a potential source of bugs and clarify the code!

coderkevin replied:
I'll add another commit to refactor this.

coderkevin commented:
@dmsnell and others, I think this code is ready to be looked at again. Please take a look at the class in `client/state/index.js` and the HOC at `client/state/use-reducer.js` especially.

mcsf commented on client/state/use-reducer.js:
```diff
@@ -0,0 +1,54 @@
+			componentWillReceiveProps( newProps ) {
+				if ( newProps.store ) {
+					removeReducer( reducerName );
+					this.store = newProps.store;
```
The `store` instance property is actively managed in this class (in `constructor` and here), yet it's seemingly never used, as `addReducer` and `removeReducer` are already bound to `state`'s `calypsoStore` anyway. At the very least, `store` will implicitly be carried in `context` or `props` into `createElement( WrappedComponent, this.props )`, but that doesn't require the setting of `this.store`. Or am I missing something?

coderkevin replied:
That's a fair point. When I wrote this, I had planned on dealing with the specific instance of the store to keep it from being bound to a global. I guess with how `client/state/index.js` ended up masking the global instance, it's not really needed here. I'll remove the instance here. Thanks!

coderkevin replied:
Oops, so looking over the code again, I see one good reason to keep this around, and it's to compare if the store was provided as a prop and has changed. Granted, that probably "should never happen" in our application, so maybe we can ignore that case. It's enough of an edge case that I'm comfortable removing support for it if we all agree.

mcsf replied:
Even better. :)

mcsf commented on client/state/index.js:
```diff
@@ -111,26 +115,113 @@ export const reducer = combineReducers( {
+export function getCombinedReducer() {
+	return calypsoStore.getCombinedReducer();
+}
+
```
This doesn't concern the overall architecture of the PR, but these could be replaced with `f = calypsoStore.f.bind( calypsoStore )` for each `f` in the set of exported functions.

coderkevin replied:
Yeah, that would be tidier here for sure. Thanks for the suggestion!

coderkevin replied:
So, I tried this out, but when I did, it caused the JSDoc comments to not generate anything anymore  about these functions. Given that these exports are the interface here, and assuming that we actually care about generating JSDoc correctly, I think we should stick with the more redundant function definitions. Further thoughts?

mcsf replied:
Hm, I suspected something like that would turn up. I guess it's a minor thing.

To be sure: the CalypsoStore class came to be in order to make unit tests easier to write, correct?

coderkevin replied:
Mostly that, also to encapsulate the code a bit, and also to avoid having a bunch of scattered globals in the file.

dmsnell replied:
JSDoc can handle it. Check out tags like [@function](http://usejsdoc.org/tags-function.html) and [@alias](http://usejsdoc.org/tags-alias.html) and [@borrows](http://usejsdoc.org/tags-borrows.html)

nb replied:
To me, the explicitness is worth the perceived duplication. Those function are the main proposed interface for developers and being able to find them defined in the source can be very useful when navigating the codebase.

mcsf commented on client/state/index.js:
```diff
@@ -111,26 +115,113 @@ export const reducer = combineReducers( {
+	getCombinedReducer() {
+		const reducers = this.reducers;
+		return combineReducers( reducers );
+	}
```
Can we memoize this function based on the reference of `this.reducers`? Since it gets called in a number of different places, repeatedly calling `combineReducers` seems avoidable.

coderkevin replied:
I thought about it, but functionally in the application it doesn't happen very often, so I resisted adding yet another piece of data to hang on to and keep current, when we're already storing these in the redux store object. I guess it's the classic tradeoff between time and space, and I figured it changes very infrequently, so I decided on saving space (however small)

mcsf commented on client/state/README.md:
```diff
@@ -17,6 +17,19 @@ const store = createReduxStore();
 All the application information and data in Calypso should go through this data flow. When you are creating a new module with new data requirements, you should add them to this global store.
 
+Alternatively, you can do the following to add a reducer at run-time within the application.
+Also, make sure you remove the reducer if you code is no longer active within the system.
```
- `s/you code/your code/` :) (also, I'd do `s/run-time/runtime/`)

- Though I personally like explicit line breaks, we usually keep one line per paragraph in Markdown documents.

coderkevin replied:
👍

mcsf commented on client/state/README.md:
````diff
@@ -17,6 +17,19 @@ const store = createReduxStore();
+
+// Call this when your code becomes inactive.
+removeReducer( 'myModule' );
+```
````
If we keep the `useReducer` HOC, it deserves a place here or somewhere.

coderkevin replied:
Just added it above this part. Thanks!

mcsf commented on client/state/use-reducer.js:
```diff
@@ -0,0 +1,54 @@
+ * @param { func } reducerFunc The function that takes (state, action) and returns state.
+ * @returns { func } A curried function that takes a component to wrap.
+ */
+export default function useReducer( reducerName, reducerFunc ) {
```
Minor, but I think the name `withReducer` would be more fitting. The `with` naming prefix is used in a number of langs (incl. JS's `with` and Python's), usually implying that there is both a set-up and a tear-down of _something_.

coderkevin replied:
Thanks for the comment. I was struggling for the right name for this HOC, so I'm definitely open to input. I'd like to hear others' opinions on this, too.

coderkevin replied:
I'd also like to note that if someone has a better idea as to where this HOC should live, I'm definitely open to that as well.

coderkevin commented:
Thanks for the helpful comments, @mcsf! I'm going to hold off for just a bit before I push an updated version to give others a chance to read the comments first before they become "outdated"

coderkevin commented:
Amended the last commit with the following changes (thanks, @mcsf!)

- Renamed `useReducer` to `withReducer` to be more descriptive.
- Removed `this.store` from the `withReducer` HOC as it was a real corner case that would ever benefit from it (and none of our current code has that corner case)
- Added example code in the README.md

dmsnell commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,30 @@
+		default:
+			return state;
+	}
+}
```
what's the purpose of storing an object here as state when it looks like all we're actually storing is a number? objects as state obscure the structure of the tree and make for complicated reducers. what if we just made initial state a primitive data type?

```js
export const lyricIndex = ( state = 0, { type } ) =>
	HELLO_DOLLY_NEXT_LYRIC === type
		? ( lyricIndex + 1 ) % lyrics.length
		: state;
```

I think it's easy to forget how simple and straightforward reducers can be. Redux works extremely well via composition (in this case with `combineReducers`) such that we can extend our state by joining together independent parts instead of adding new things into old things (storing objects as state and adding new properties).

For example, in `advanceToNextLyric()` it looks like we're leaving the door open to new future properties of this state object, but I would ask why we need or want that. If we have a new and different piece of data, it should get its own reducer entirely.

> Redux pushes us to think "what changes this data?" instead of "what data does this change?"

coderkevin replied:
Thanks for the tip here. I'm not used to using tree-based `combineReducers()` calls. If it can simplify our redux state handling, then I'm all for it.

dmsnell commented on client/extensions/hello-dolly/state/selectors.js:
```diff
@@ -0,0 +1,11 @@
+export function getCurrentLyric( reduxState ) {
+	const state = reduxState.helloDolly;
+
+	return lyrics[ state.lyricIndex ];
```
just a quick note on this example - it's open to type errors if `reduxState.helloDolly` isn't available (`state.lyricIndex`) and range errors if the index is longer than the lyrics array. I'm a big fan of using `lodash.get()` with a default value which we know if _always_ safe to use, but if we don't want to pull in `lodash` for some reason it might be wise to guard ourselves here.

```js
export const getCurrentLyric = state => {
	// default to first lyric if unavailable
	const index = get( state, 'helloDolly.lyricIndex', 0 );

	// wrap around if given too high of an index
	return lyrics[ index % lyrics.length ];
```

you may notice that I made some decisions in this code that could be decided differently: what happens when the state fails us? here I decided that we should simply wrap back to the beginning. maybe an empty string or `null` would have also made good return values.

dmsnell replied:
another note, especially given that this is example code for others, is a use of dependency injection that I use in order to make testing much easier. notice how we cannot test or call `getCurrentLyric()` without the hard coupling to the lyrics array…

```js
export const getCurrentLyric = lyrics => state => {
	…
}

export default getCurrentLyric( lyrics );
```

in this code the function takes the array of lyrics as its first curried argument and returns the new and _more real_ selector as a result. now if we want to test the function we can use simple data mocks…

```js
import { getCurrentLyric } from '…';

describe( '#getCurrentLyric', () => {
	const threeSongs = [ 'one', 'two', 'three' ];
	const dollyState = state => ( { helloDolly: state } )

	it( 'should return the first item if no state index exists', () => {
		expect( getCurrentLyric( threeSongs )( dollyState( {} ) ).to.equal( 'one' );
	} );

	it( 'should wrap-around in the list if given out-of-bounds indices', () => {
		expect( getCurrentLyric( threeSongs )( dollyState( { lyricIndex: 2 } ).to.equal( 'three' );
		expect( getCurrentLyric( threeSongs )( dollyState( { lyricIndex: 3 } ).to.equal( 'one' );
		expect( getCurrentLyric( threeSongs )( dollyState( { lyricIndex: 4 } ).to.equal( 'two' );
	} );
} );
```

and now we have a couple tests which expressly test the _behavior_ of the function

coderkevin replied:
This makes sense, I've just started getting in the habit of currying bits of data along the way.

dmsnell commented on client/extensions/hello-dolly/state/test/reducer.js:
```diff
@@ -0,0 +1,28 @@
+		const state = reducer( undefined, { type: undefined } );
+
+		expect( state.lyricIndex ).to.be.a( 'number' );
+	} );
```
I'll just poke here and point out that because of how we used an object as our state instead of a primitive data type it made us feel like we needed to test this. however, if we stick to the simpler reducers and then combine them with `combineReducers()` (when needing more than one property at a given subtree) then these tests have already been rigorously made and vetted in Redux.

in this sense we end up less with a unit test here and more with an integration test or feel-good test telling us we didn't botch something up

notice also that because we have jumped into the fringe of what unit testing means for us we have designed something to test the inner workings of how our function works and have overlooked the _behavior_ which the unit test is designed to assert - we've dug _into_ the black box instead of guarantee its interface

```js
it( 'should initialize to the first index', () => {
	expect( reducer( undefined, { type: '@@test/INIT' } ).lyricIndex ).to.equal( 0 );
} );
```

I also included a string value for the action type because I believe that there can be problems with some reducers when given `undefined`

dmsnell commented on client/state/index.js:
```diff
@@ -111,26 +115,113 @@ export const reducer = combineReducers( {
+	}
+
+	removeReducer( name ) {
+		const { [ name ]: removedReducer, ...remainingReducers } = this.reducers;
```
nice use of destructuring

dmsnell commented on client/state/index.js:
```diff
@@ -111,26 +115,113 @@ export const reducer = combineReducers( {
+	}
 }
+
+const calypsoStore = new CalypsoStore();
```
what's the value we gain when we turn the store into a stateful class vs keeping it as a simpler object?

I'm not entirely opposed, but I wonder if we have overlooked a way that's simpler and easier to test and less prone to bugs by keeping the store as the variable `calypsoStore` and creating functions which return entirely new stores (which CalypsoStore.createReduxStore() actually does already).

maybe it would be significantly easier to test functions like `addReducers()` if we could test them in complete isolation and use just a basic data mock for the store.

```js
addReducer( prevStore )( name, newReducer ) -> nextStore

// or

addReducer( prevStore, name, newReducer ) -> nextStore
```

something like that ^^^

coderkevin replied:
An approach like this would require the calling code to have access to the store, which is probably not the case. The calling code would be doing this upon initialization (i.e. creation of the React components) at which time the store may not have been created. And even if it was, your component wouldn't yet have access to it until it's rendered.

This doesn't affect you if you're using the HOC, as it takes care of that for you. But if you're trying to do it yourself, it does.

dmsnell commented on client/state/with-reducer.js:
```diff
@@ -0,0 +1,45 @@
+
+			render() {
+				const result = createElement( WrappedComponent, this.props );
+				return result;
```
any reason to create `result` here instead of just…

```js
render() {
	return createElement( WrappedComponent, this.props );
}
```

coderkevin replied:
Not really, it's basically an artifact of code that used to be there.

dmsnell commented:
I'm very glad to see you working on this @coderkevin - I'm excited to be getting a good example plugin in Calypso.

While I'm pretty happy with your code (despite my flurry of comments) there's one thing that keeps bugging me about `helloDolly` and I'm reluctant to bring it up - probably because I don't see an easy answer to it. The data model we have chosen invites a situation where valid-looking data is actually invalid state. I've really taken to the expression, "making impossible states impossible" (see Richard Feldman's [talk on the subject](https://www.youtube.com/watch?v=IcgmSRJHu_8) for inspiration). What does it mean if we have `lyricIndex = -1`? What if we have `lyricIndex = 200` and only ten lyrics? What if we have `lyricIndex = 3` but the list of lyrics change?

By creating an indirect reference to an item in the list of lyrics we've decoupled what is intrinsically coupled: which item in a list we have selected…

Let me illustrate with a naive approach which makes for universally valid state but is much more memory hungry:

```js
import { lyrics } from './constants'

const helloDolly = ( state = lyrics, { type } ) =>
	HELLO_DOLLY_NEXT_LYRIC === type
		? [ ...lyrics.slice( 1 ), ...lyrics.slice( 0, 1 ) ]
		: state
```

in this case we have completely eliminated the index! on each update we simply rotate the list of lyrics and our selector can become `get( state, 'helloDolly[ 0 ]', 'no song so sad' )`. now it's obviously heavy because we end up storing the whole list of lyrics in memory in the state tree. maybe it's not that bad though since our list is small and this may not get called too often (otherwise we might get garbage-collection thrashing). notice too that as long as `lyrics` is _any_ list (even an empty list) then we have no risk of errors. the list functions are safe on empty lists.

we could probably tradeoff memory usage for processing speed by using a variant which finds the index of the current lyric in the array and then grabs the next one.

the point of this example is that I think we may not have exhausted our search for the right structure of our data. it's very possible that by changing the way we store and reference the lyrics that our logic could dramatically simplify, grow in resilience to bugs, and make testing easier too. Maybe we could inspire a few people along the way :smile:

dmsnell commented:
another high-level comment I'll poke at is that we're demonstrating creating new action types and calling for specific data changes instead of app changes. this is odd because `helloDolly` is an odd plugin. however, the important question Redux pushes us to ask is _What changes this data?_ so I will post the question…

> When should the Hello Dolly lyrics update?

personally I'd prefer to never see `dispatch( { type: HELLO_DOLLY_NEXT_LYRIC } )` because that doesn't tell me anything about what intentions the user was communicating when the lyrics updated.

suppose we wanted to answer the question with "the lyrics should update on every 'page' load". let's look into `state/action-types.js` and see if we have an existing action which conveys that notion. well, I admit, we don't have a solid navigation and routing system in Redux in Calypso, but we have a few actions like `ROUTE_SET` or `SECTION_SET`

maybe we also want to update the lyric every time the user saves some site setting: `SITE_SETTINGS_SAVE`

if we use _action_-describing action types in our reducers we can hint at what kinds of things should be affecting our data, where to look when we wonder why it changed, and reduce code bloat. in other words, if we create a new action type to change each data type then we end up with a very large system which looks like a big class with property setters. in fact, what we want is a collection of independent data types taking responsibility for themselves to respond to changes in the app itself.

```js
export const selectedLyric = ( state = …, { type } ) => {
	switch ( type ) {
		case ROUTE_SET:
		case SECTION_SET:
		case SITE_SETTINGS_SAVE:
			return nextLyric( state );

		default:
			return state;
	}
};
```

so how here do we go about adding a new transition to the lyrics? we go to a single place in a single file - the data's reducer - _not_ to ten different files where we have called `NEXT_LYRIC`

hope this is helpful!

gziolo commented on client/state/with-reducer.js:
```diff
@@ -0,0 +1,45 @@
+			constructor( props, context ) {
+				super( props, context );
+				this.reducerName = reducerName;
+				this.reducerFunc = reducerFunc;
```
Those two instance variables are never used.

coderkevin replied:
Correct. This is left over from other code I had since removed. Thanks!

coderkevin commented:
>well, I admit, we don't have a solid navigation and routing system in Redux in Calypso, but we have a few actions like ROUTE_SET or SECTION_SET

This is precisely what makes me very nervous about attaching directly to existing Calypso actions. Right now, we have no contract of any kind that says these actions won't change. So the way the actions work, when they're fired, the data they carry, and even their very existence is subject to change. This is not a solid foundation upon which to build an extension to any application.

I think fairly soon, we're going to need to have some further conversations about how we can "officially" support extensions in a solid way where they can have some measure of confidence that they're not built upon shifting sand.

dmsnell commented:
> This is precisely what makes me very nervous about attaching directly to existing Calypso actions. Right now, we have no contract of any kind that says these actions won't change. So the way the actions work, when they're fired, the data they carry, and even their very existence is subject to change. This is not a solid foundation upon which to build an extension to any application.
> 
> I think fairly soon, we're going to need to have some further conversations about how we can "officially" support extensions in a solid way where they can have some measure of confidence that they're not built upon shifting sand.

those actions won't likely go away. the trend will probably be that we'll use them more thoroughly and eventually we'll base routing off of them.

mtias commented:
> Add UI component to Hello Dolly example

@coderkevin what does this mean specifically? Hello Dolly is already using a bunch of UI components.

coderkevin commented:
@mtias It's shorthand for adding a UI component to show the lyrics of the song using Redux state. I'll update the checklist to be more descriptive here.

dmsnell commented on client/extensions/hello-dolly/state/reducer.js:
```diff
@@ -0,0 +1,23 @@
+
+function advanceToNextLyric( index ) {
+	return index + 1;
+}
```
at this point with the much simplified code, do we even need this function? seems simple enough to inline the logic above at `state + 1`

coderkevin replied:
I think there's value in having a function here as it makes the code's intent more obvious. If we didn't have a function call here, we'd need a comment to explain why we're doing `index + 1`. Unless there's an important performance reason, I'd like to keep the function for this.

dmsnell replied:
fair enough. with _such simple logic_ it's harder for me to see it this way, but make the logic just a little more complicated and I'm 💯 with you.

dmsnell replied:
@coderkevin I think part of the reason you feel like keeping this function makes the intent more clear could be related to the fact that the reducer has no name. `advanceToNextLyric()` seems like a function which updates some internal reference to a lyric _but this reducer itself is that reference_.

while I won't press you to reconsider, I might suggest that you give the reducer a fitting name. that in itself might clarify the logic enough and maybe you wouldn't see the same demand for this function.

let me rephrase: the reducer isn't a function which is responsible for all of the lyric state - it's a simple function governing a tiny piece of primitive data. _any_ and _all_ reducers governing the state of some system or subsystem are _a composition of little reducers_.

```js
export const lyricIndex = ( state = 0, { type } ) => { … };

// or in the style you've written…

export function lyricIndex( state = 0, { type } ) {
	…
}
```

for now that's good enough to encapsulate all state. if we had a need for more or we knew we would expand it in the future we use `combineReducers()`…

```js
export default combineReducers( {
	lyricIndex,
} );
```

coderkevin replied:
Added the reducer name, like we discussed. Thanks!

dmsnell commented on client/extensions/hello-dolly/state/test/reducer.js:
```diff
@@ -0,0 +1,27 @@
+		expect( reducer( 13, { type: SITE_SETTINGS_SAVE } ) ).to.equal( 14 );
+	} );
+} );
+
```
nice tests!

dmsnell commented on client/extensions/hello-dolly/state/test/selectors.js:
```diff
@@ -0,0 +1,27 @@
+			expect( getCurrentLyric( threeLines )( dollyState( 3 ) ) ).to.equal( 'one' );
+			expect( getCurrentLyric( threeLines )( dollyState( 4 ) ) ).to.equal( 'two' );
+		} );
+	} );
```
what if our lyrics array is empty?

coderkevin replied:
Completely right. Thanks! I added a test and code to support it.

coderkevin commented:
@dmsnell, @mtias, @nb, @aduth, @ehg, this PR should be complete at this point. Can I get reviews/final comments on this so we can get it merged?

Thanks!

justinshreve commented:
PR builds successfully for me and works correctly at `/hello-dolly`!

dmsnell commented on client/state/test/index.js:
```diff
@@ -12,7 +12,39 @@ import { useSandbox } from 'test/helpers/use-sinon';
+		it( 'should add and remove a reducer by name', () => {
+			const store = createReduxStore();
+
+			should.not.exist( store.getReducers()[ name ] );
```
why import `expect` and then also use `should`? is there a reason to include both?

```js
expect( store.getReducers()[ name ] ).to.not.exist;
```

coderkevin replied:
Because I started out using `should`, and then used `expect` by your recommendation. ;) I'll consolidate.

dmsnell commented on client/extensions/hello-dolly/hello-dolly-page.js:
```diff
@@ -0,0 +1,101 @@
+import lyrics from './state/lyrics';
+import { nextLyric } from './state/actions';
+
+class HelloDollyPage extends Component {
```
sorry for coming in late here on this (or for missing it) but we strongly encourage a "stateless-functional component-first" approach that only switches into a `Component` once a need for the lifecycle methods or `state` has been established. neither requirement seems to have been met here so I'm wondering why this is extending `Component` instead of being a simpler render-function.

coderkevin replied:
The whole point of this PR is to establish redux state in extension code, using hello-dolly as an example. So I'm not sure how you feel state doesn't belong in this component when it needs to both display the current lyric and also dispatch an action. I suppose we could pass it up the chain to the index, but does it really make sense there? The index doesn't care about lyrics. This component does.

dmsnell commented on client/lib/signup/test/dependency-store.js:
```diff
@@ -23,7 +23,7 @@ describe( 'dependency-store', function() {
 		SignupActions = require( '../actions' );
 
-		const store = createStore( reducer );
+		const store = createStore( combineReducers( initialReducers ) );
```
why are we changing these thins here instead of where `initialReducers` is defined? it seems like a leak in design to force the dependents of the code to update to reflect what appears as internal changes…

coderkevin replied:
These are test functions, and probably shouldn't even be bringing in all initalReducers anyway. They should probably only add the reducer they care about to their test store. I was just trying not to rewrite everything while implementing some changes. 😛  I figure this PR is big enough already.

dmsnell commented on client/state/README.md:
```diff
@@ -17,6 +17,26 @@ const store = createReduxStore();
+store.addReducer( 'myModule', myReducerFunction );
+
+// Call this when your code becomes inactive.
+store.removeReducer( 'myModule' );
```
these sound like a dangerous set of operations to expose. do we have a need for plugins to use these over the `withReducer()` wrapper?

coderkevin replied:
If you wanted to implement your own lifecycle, this would be how you would do it. As @mcsf pointed out below, we may want to incorporate reducer lifecycle into something more sophisticated at the extension level at some point in the near future. It's also used for unit testing.

coderkevin replied:
Also, without this code, `replaceReducer` is available, which is far more dangerous than either of these function. Isn't this one step *less* dangerous anyway?

dmsnell replied:
`replaceReducer` is exposed as part of the store API and might be necessary for store enhancers, but shouldn't make it past a code review if someone tries to call it inappropriately. on the other hand, I don't actually have a problem hiding it altogether from everything past the core Redux subsystem.

the difference here I believe is the fact that we're putting in code in a guide on using `addReducer()` and `removeReducer()` and those things seem straightforward and _blessed_ by their inclusion. if we hide those by default and make people make active changes to use them we might head off inappropriate uses.

just asking the question: is this necessary? is there _any_ reasonable case we can imagine needing it? could we get away with keeping it out?

coderkevin replied:
That's a fair point on the documentation. I think you're right in that we shouldn't encourage people to use these interfaces. We'll still need the functions to exist, in order for `withReducer` to use them, but we don't have to advertise that they're there to be used. I'll remove the mention of them from the README.md

dmsnell replied:
might be worth putting a note in the docblock comments indicating that too

```js
/**
 * Merges in a custom reducer with a
 * custom name into the state tree.
 *
 * @internal
 *
 * NOTE: You probably shouldn't be using this function.
 * please make sure that if you do you _really_ need to
 * and that you get extra code review before merging
 * in any changes.
 *
 * @see withReducer
 */ 
```

dmsnell commented on client/state/index.js:
```diff
@@ -111,9 +115,16 @@ export const reducer = combineReducers( {
+ *
+ * @return { array } The array of enhancers to be configured in a redux store.
+ */
+export function createEnhancers() {
```
create? how about `getEnhancers()` or something that doesn't seem to communicate that we are creating new ones instead of simply returning a list of existing ones?

coderkevin replied:
It does create a list of enhancers though. What it returns does not exist anywhere until it's created by this function.

dmsnell replied:
okay. that doesn't match what I think of in `create` but we're allowed to disagree.

this function doesn't generate any enhancer functions whose behavior is dependent on runtime circumstances. it merely returns a list of enhancers which are generated in other files…

dmsnell commented on client/state/index.js:
```diff
@@ -132,5 +143,75 @@ export function createReduxStore( initialState = {} ) {
 }
+
+/**
+ * Named export function for creating a redux store.
```
"Named export function" seems a bit redundant here

coderkevin replied:
Fair enough. Reworded to be more descriptive.

dmsnell commented on client/state/index.js:
```diff
@@ -132,5 +143,75 @@ export function createReduxStore( initialState = {} ) {
+	function addReducer( name, func ) {
+		if ( reduxStore.reducers[ name ] !== undefined ) {
+			throw new Error( `addReducer(): Reducer by name "${ name }" already in use` );
+		}
```
seems like we would want to be able to safely call `addReducer` from multiple components with the same reducer. what if we have the same button instantiated in two places?

couldn't we made this idempotent on the combination of name and function?

```js
const existingReducer = reduxStore.reducers[ name ];

if ( existingReducer !== undefined && existingReducer !== func ) {
	throw new Error( … );
}
```

coderkevin replied:
I thought about this before, but it really does open a new can of worms. If we have two pieces of code call `addReducer`, presumably they might also call `removeReducer`. What happens when code A yanks the reducer out of the chain and code B still needs it?

In order for it all to work this way, we will have to disallow reducers to be removed from the chain, and do we really want that? I feel it will catch up with us eventually and we're already pushing the limits of memory and time delays.

dmsnell replied:
another quandary which anticipates trouble with exposing these two functions…

coderkevin replied:
I think we could eventually implement a dependency sort of approach to this, where code declares its dependence on a reducer and the reducer is added/removed based on that. But I see that as a layer on top of `addReducer` and `removeReducer` which would be in a future PR. But that would require holding more state as well.

dmsnell commented on client/state/with-reducer.js:
```diff
@@ -0,0 +1,65 @@
+			constructor( props, context ) {
+				super( props, context );
+				this.store = props.store || context.store;
+				this.propsMode = Boolean( props.store );
```
what is this for? it will only ever validly be `true` because otherwise we'll have an error raised below… do we even use it later on?

coderkevin replied:
To my knowledge, no component in Calyspo uses props for this, but it is an accepted Redux pattern. In fact, I patterned this code off of the `react-redux` `connect()` code: https://github.com/reactjs/react-redux/blob/master/src/components/connectAdvanced.js#L122

dmsnell replied:
I'm asking about `this.propsMode` just so you know - not `this.store = props.store || context.store`

coderkevin replied:
Oh, gotcha. Yeah, I guess this isn't needed anymore. I made some simplifications to the code that rendered it obsolete. I'll remove it. Thanks!

dmsnell commented on client/state/index.js:
```diff
@@ -132,5 +143,75 @@ export function createReduxStore( initialState = {} ) {
+		removeReducer,
+		getReducers,
+	};
+}
```
@coderkevin I'm still a bit bothered by the statefulness of this. did you consider an approach I have often used which allows for the state fulness of data while preserving the independence and testability of the functions?

```js
let reducers = {};
let reducer = identity;

export const replaceReducers = newReducers => ( {
	reducers = newReducers;
	reducer = combineReducers( newReducers );
} );

export const addReducerTo = ( existingReducers, name, newReducer ) => ( {
	...existingReducers,
	[ name ]: newReducer,
} );

export const addReducer = ( name, newReducer ) =>
	replaceReducers( addReducerTo( reducers, name, newReducer ) );

export const removeReducerFrom = ( existingReducers, name ) => {
	const { [ name ]: droppedReducer, ...remainingReducers } = existingReducers;

	return remainingReducers;
}

export const removeReducer = name =>
	replaceReducers( removeReducerFrom( reducers, name ) );
```

coderkevin replied:
In this example, where is `existingReducers` held? There are two data members that have to be held in lock-step for this to work. One is the redux store (so `replaceReducer` can be called upon it), and the other is the list of existing reducers (so the new combined reducer can be generated upon add/remove of a reducer.) These two pieces of state must exist together at all times. That's why I added the list of reducers to the store itself. Otherwise, the responsibility of keeping them together just gets pushed down the line to the calling code, and that's what I feel we should avoid.

The testability I believe is preserved by having no global state, and allowing test code to create as many redux stores as it wishes. It's worked so far.

dmsnell replied:
> That's why I added the list of reducers to the store itself. Otherwise, the responsibility of keeping them together just gets pushed down the line to the calling code, and that's what I feel we should avoid.

sure - that works! the point is getting them out as open as possible - with as few actual dependencies as is possible.

> having no global state

but we're still dealing with considerable shared mutable state when we start manipulating `this` - if we can return entirely new objects when mutating (preserving the old ones) then we can free ourselves from that shared state entirely

coderkevin replied:
I don't think we can return an entirely new store each time the reducers list changes, can we? We'd have to update every `<Provider>` out there to do that and we don't have a centralized list of those. So we're stuck with mutating the redux store at least, every time a reducer is added or removed. I guess I don't see adding a reducers array to that store as "considerable shared mutable state", I see it as just what's necessary to make this work. Do you have a better idea of how to keep these things linked together? (the redux store and the reducers array)

dmsnell replied:
> I don't see adding a reducers array to that store as "considerable shared mutable state"

I have no problem with storing the reducers array to the store. the shared state part was about unhinging existing code which held a reference to a store that unexpectedly changes.

what did you think of the code I posted here: separating the stateful from the stateless by creating testable functions which return new new things and then exported functions which get called and actually mutate the existing store with those new things (the reducers in this case)



> We'd have to update every <Provider> out there

you bring up a good point. that could be a very critical roadblock for removing that statefulness.

coderkevin replied:
I like the separation of the two types of functions, but I do wonder about the use of module globals in `replaceReducers`. I guess I've been on the fence about module globals versus object instance variables. I have a long history in many languages that tells me globals aren't great, even in a module. But I've also noticed they are a bit more accepted in JS than other languages as well.

dmsnell replied:
well the instance variables are still globals, they are just globals hidden in other globals :wink: you seemed to indicate that we couldn't reasonably create two instances of this class which makes me think we're creating a class for a single global data structure and only ever creating one instance of that class…hopefully you can see where I'm going with this

dmsnell commented:
more in-depth review of the framework code :+1:

kellychoffman commented:
Fixed the `Next` button alignment in 597ed13 after a convo with @coderkevin. 

Note for the future, as its quite out of scope for this PR: I'd like to come up with a better solution for css that isn't inline styles.

coderkevin commented on client/extensions/hello-dolly/state/selectors.js:
```diff
@@ -0,0 +1,9 @@
+
+	return lyrics[ index % lyrics.length ] || translate( 'I can\'t think of a song to sing.' );
+};
+
```
Sounds reasonable.

coderkevin commented:
@mcsf I completely agree that a more sophisticated solution could be in store (no pun intended!) for this code. I think a more comprehensive, robust extension system would benefit a lot of folks trying to accomplish what we are starting to do here. However, I've refrained from that approach just yet for a few reasons:

1. We're just getting started with this, we don't know the exact shape the extension system will take, so it's best to make as few assumptions as possible until we know more.
2. The rest of the extension system that would be expected to support such a thing does not yet exist.
3. It would be a lot of code. @matticbot already categorizes this PR as XL. I just wanted to get redux state available in extension code with the least amount of reasonable code possible.

dmsnell commented:
this is just a reminder that this PR will need HEAVY testing and review before we merge due to the fact that it fundamentally changes the way our entire state system 

 - [ ] Tests in Docker
 - [ ] Clear the browser caches `localStorage.clear(); indexedDB.deleteDatabase( 'calypso' )`
 - [ ] Sign in outside of the proxy and create a user and a site
 - [ ] Write and publish a post
 - [ ] Test on staging after merge and before deploy

This is nothing personal against you @coderkevin at all - it's just an attempt at putting together a deploy checklist for critical code. @alisterscott or @lancewillett may have some insight to further things to checkoff before we hit the big 🔴.

coderkevin commented:
@dmsnell Agreed. It does basically affect everything to do with the redux state, which in turn affects all of Calypso, basically. On the bright side, I think the chances of subtle issues is fairly low, as if anything fails it all goes down in flames. 😆

coderkevin commented:
@dmsnell I added your checklist to the top-level PR comment so we can check them off. If anyone has something to add, feel free to add it there and just comment below as to the reasoning, if needed.

allendav commented on client/extensions/hello-dolly/hello-dolly-page.js:
```diff
@@ -0,0 +1,101 @@
+import SectionHeader from 'components/section-header';
+import FAQ from 'components/faq';
+import FAQItem from 'components/faq/faq-item';
+import getLyric from './state/selectors';
```
I find it helps to use the same name on the `import` as the module exports - i.e. it helps when searching the code base and it keeps one less than on the programmer's mental stack. Would you mind changing `getLyric` to `getCurrentLyric`?

coderkevin commented:
This PR is being deferred in favor of a more static approach for now (#11728). However, the dynamic approach will need to be revisited at some point in the future. Issue #12582 was created to track the future implementation of this enhancement.

Note: I'd like to keep this PR open until it is implemented, so as to not lose the code that exists on this branch until it has been successfully re-implemented.

dmsnell commented:
@coderkevin can we close this PR? note that even after deleting the branch in GitHub we can reopen it and the code will live on.

coderkevin commented:
We've got enough mention of it, I think we won't lose track of it. I'll close it.
